File: Languages/cpp/algorithms-dsa/leetcode/sliding-window/longest_repeating_char_replacement.cpp
/*
 * LeetCode 424 - Longest Repeating Character Replacement
 * Topic: Sliding Window
 * Difficulty: Medium
 *
 * Sliding window: track frequency of each char. Window is valid if
 * (window_size - maxFreq) <= k.
 * Time: O(n)
 * Space: O(1) (26 letters)
 */
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
using namespace std;

int characterReplacement(string s, int k) {
    vector<int> count(26, 0);
    int left = 0, maxFreq = 0, result = 0;
    for (int right = 0; right < (int)s.size(); right++) {
        count[s[right] - 'A']++;
        maxFreq = max(maxFreq, count[s[right] - 'A']);
        while ((right - left + 1) - maxFreq > k) {
            count[s[left] - 'A']--;
            left++;
        }
        result = max(result, right - left + 1);
    }
    return result;
}

int main() {
    assert(characterReplacement("ABAB", 2) == 4);
    assert(characterReplacement("AABABBA", 1) == 4);
    assert(characterReplacement("AAAA", 0) == 4);
    assert(characterReplacement("ABCD", 2) == 3);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/sliding-window/longest_repeating_char_replacement.py
"""
LeetCode #424 - Longest Repeating Character Replacement
Topic: Sliding Window
Difficulty: Medium

Find longest substring with at most k replacements using window + max frequency.

Time Complexity: O(n)
Space Complexity: O(1) - at most 26 keys
"""


def character_replacement(s: str, k: int) -> int:
    counts: dict[str, int] = {}
    left = 0
    max_freq = 0
    result = 0
    for right in range(len(s)):
        counts[s[right]] = counts.get(s[right], 0) + 1
        max_freq = max(max_freq, counts[s[right]])
        while (right - left + 1) - max_freq > k:
            counts[s[left]] -= 1
            left += 1
        result = max(result, right - left + 1)
    return result


if __name__ == "__main__":
    assert character_replacement("ABAB", 2) == 4
    assert character_replacement("AABABBA", 1) == 4
    assert character_replacement("A", 0) == 1
    print("All tests passed!")
